src/main/java/org/jenkinsci/plugins/workflow/log/FileLogStorage.java

```diff
@@ -0,0 +1,276 @@
+                                }
+                                lastId = space == -1 ? null : line.substring(space + 1);
+                            }
+                            if (pos == lastTransition) {
+                                if (hadLastId) {
+                                    w.write(LogStorage.endStep());
+                                }
+                                hadLastId = lastId != null;
+                                if (lastId != null) {
+                                    w.write(LogStorage.startStep(lastId));
+                                }
+                            }
+                            super.write(b);
+                            pos++;
+                        }
+                        @Override public void flush() throws IOException {
+                            if (lastId != null) {
+                                w.write(LogStorage.endStep());
+                            }
+                            super.flush();
+                        }
+                    });
+                    ConsoleAnnotators.setAnnotator(caos.getConsoleAnnotator());
+                    return r;
+                }
+            }
+        };
+    }
+
+    @Override public AnnotatedLargeText<FlowNode> stepLog(FlowNode node, boolean complete) {
+        String id = node.getId();
+        try (ByteBuffer buf = new ByteBuffer();
+             RandomAccessFile raf = new RandomAccessFile(log, "r");
+             BufferedReader indexBR = index.isFile() ? Files.newBufferedReader(index.toPath(), StandardCharsets.UTF_8) : new BufferedReader(new NullReader(0))) {
+            String line;
+            long pos = -1; // -1 if not currently in this node, start position if we are
+            while ((line = indexBR.readLine()) != null) {
+                int space = line.indexOf(' ');
+                long lastTransition = -1;
+                try {
+                    lastTransition = Long.parseLong(space == -1 ? line : line.substring(0, space));
+                } catch (NumberFormatException x) {
+                    break; // corrupt index file; forget it
+                }
+                if (pos == -1) {
+                    if (space != -1 && line.substring(space + 1).equals(id)) {
+                        pos = lastTransition;
+                    }
+                } else {
+                    raf.seek(pos);
+                    if (lastTransition > pos + Integer.MAX_VALUE) {
+                        throw new IOException("Cannot read more than 2Gib at a time"); // ByteBuffer does not support it anyway
+                    }
+                    // TODO can probably be done a bit more efficiently with FileChannel methods
+                    byte[] data = new byte[(int) (lastTransition - pos)];
+                    raf.readFully(data);
+                    buf.write(data);
+                    pos = -1;
+                }
+            }
+            if (pos != -1) {
```

Comment:
Comment: // Reading out the final dangling chunk of log content if we're still writing content for the node.